kernel/proc.c

```diff
@@ -146,29 +180,73 @@ allocproc(void)
   p->context.ra = (uint64)forkret;
   p->context.sp = p->kstack + PGSIZE;
 
+  // Some other fields
+  p->killed = 0;
+  p->chan = 0;
+  p->xstate = 0;
+
+  // Acquire head and head->next locks if head->next exists
+  int at_least_two = proc_list->next != proc_list;
+  acquire(&proc_list->cur_proc->lock);
+  struct proc_list *proc_next = proc_list->next;
+  if (at_least_two) acquire(&proc_next->cur_proc->lock);
+  proc_lst_push(proc_list, cur_list);
+  if(at_least_two) release(&proc_next->cur_proc->lock);
+  release(&proc_list->cur_proc->lock);
   return p;
 }
 
+// Must hold write_lock to call
+void remove_unused_from_proc_list() {
+  if (!rwlock.write_lock.locked)
+    panic("Remove something from proc list without write lock");
+
+  struct proc_list *cur;
+  for (cur = proc_list->next; cur != proc_list; cur = cur->next) {
+    if (cur->cur_proc->state != UNUSED) continue;
+
+    // There always is prev, at least head
+    cur = cur->prev;
+    struct proc_list *removed = cur->next;
+    proc_lst_remove(removed);
+    kfree(removed->cur_proc);
+    kfree(removed);
```

Comment:
You are a bit confused about the flow of the program. In freeproc we set proc->state = UNUSED and THEN we get into that function and actually free the memory. So calling freeproc from here is pointless since we already been there